### DISCUSSION TITLE
Reach consensus on push/get/put order

### DISCUSSION THREAD

**Discussion by Izaakwltn** [coalton-lang/coalton#996] [Polls]
Referring to Issue #938, I think we should try to reach consensus on the order of arguments in push, get, and put functions.

**Comment by stylewarning** [comment]
Thought: Currying the (quasi-)mutable collection seems wrong to me.

**Reply by stylewarning** [reply]
Elaboration: I think thinking about what you want to do with the curried versions informs us of what reads well. `(nth n)` is a good function to get the nth of something, while `(ref dict)` (a function which takes a key and returns a value) and its companion `(get "5")` (a function that takes a collection and returns a value) read nice in their respective cases.

**Comment by Izaakwltn** [comment]
I've also just had a moment of doubt, after recalling that the standard for Lisp list operations is arguments first:
`(nth 0 ls)`
`(push obj objs)`
`(sort #'foo barlist)`

(Of course, lisp arrays use the reverse order so that may be a weak point...)

**Comment by macrologist** [comment]
When and if loops become available, there would be less of a use-case for curried destructive functions. 

At the present time, ``` (iter:foreach (vector:push vec) (map g (iter:up-to some-number)))``` is kind of natural. 

With loops, it becomes something like (perhaps)

```

(for i in (iter:up-to some-number) (vector:push vec (g i)))

```

**Reply by stylewarning** [reply]
I'd be happy if we just agreed on an accessor argument ordering to start with, and find a convention on the name. I think we should have two accessor names: GET and REF, with the key-or-index as the first and second arguments respectively in each.

**Reply by macrologist** [reply]
Works for me. Both terms are semantically broad enough to cover whatever exotic index types we happen to find ourselves passing.

**Reply by stylewarning** [reply]
> Works for me. Both terms are semantically broad enough to cover whatever exotic index types we happen to find ourselves passing.

I also abhor using FLIP (or any other permuting combinator) for such common operations.

**Reply by macrologist** [reply]
I've been using flip - its fine about half the time; illegible the other half.
